UVES Popler support for coadding data

rcooke-ast opened the pull request:
I have added support to combine PypeIt data products with UVES_popler. The [UVES_popler changes](https://github.com/MTMurphy77/UVES_popler/pull/8) have now been merged, although there may still be a few minor issues that need to be sorted out. Michael Murphy has kindly joined the User's Slack and is happy to assist users with some support if they choose to use this package. I've also updated some of the docs.

Overall, if this is something that receives a lot of interest, we may want to consider redeveloping a GUI for such data combination in the future, possibly with python so users don't need to install CFITSIO and PGPLOT...

rcooke-ast committed:
added DATETIME

rcooke-ast committed:
rm TODO

rcooke-ast committed:
UVES_popler support

rcooke-ast committed:
add blaze

rcooke-ast committed:
add blaze function calculation

rcooke-ast committed:
add blaze function calculation

rcooke-ast committed:
Merge branch 'develop' into uves_popler

# Conflicts:
#	doc/releases/1.15.1dev.rst
#	pypeit/core/skysub.py

rcooke-ast committed:
add docs

rcooke-ast committed:
updates

kbwestfall reviewed (commented):
This looks fine to me, but I think we need reviews from @debora-pe and @jhennawi .  Two requests from me:

 - Increment the datamodel version number for `SpecObj`.
 - I dislike our use of the term "blaze function" here.  The flat-field observations do not strictly provide the blaze function only (unless I'm missing something).  Maybe that's the dominant term, but there are other components (e.g., the detector QE, etc).  Why not just use "FLAT" instead?

debora-pe commented:
It seems a good idea to have the "blaze function" in the SpecObj, but we may want to consolidate this with what `pypeit_coadd_1dspec` is doing, which also uses the blaze. @jhennawi what do you think?

Also, @rcooke-ast can you add a quick example of how to use UVES_popler with PypeIt spec1ds?

rcooke-ast committed:
update docs

rcooke-ast committed:
change blaze to flat

rcooke-ast commented:
Thanks for the quick feedback @kbwestfall and @debora-pe! Responding to your queries:

> Increment the datamodel version number for SpecObj.

Done!

> I dislike our use of the term "blaze function" here. The flat-field observations do not strictly provide the blaze function only (unless I'm missing something). Maybe that's the dominant term, but there are other components (e.g., the detector QE, etc). Why not just use "FLAT" instead?

You're right, it's not strictly the blaze, but this is how UVES_popler reads in the PypeIt files. I have changed this to FLAT, and I'll submit a small PR to UVES_popler.

> Also, @rcooke-ast can you add a quick example of how to use UVES_popler with PypeIt spec1ds?

I've linked to the documentation, and I don't wish to reproduce too much here, but I've added some of the first simple steps to get things going. For the record, here is the documentation for your interest: 

[https://astronomy.swin.edu.au/~mmurphy/UVES_popler/](https://astronomy.swin.edu.au/~mmurphy/UVES_popler/)

Also, Michael Murphy has a video tutorial of how to interact with UVES_popler, and is going to share that with us.

debora-pe commented on doc/coadd1d.rst:
```diff
@@ -303,3 +303,45 @@ and launches a GUI from the `linetools`_ package. e.g.:
+
+For details on how to use the tool, please refer to the
+`UVES_popler documentation <https://astronomy.swin.edu.au/~mmurphy/UVES_popler/>`__.
+To get you started with reading in PypeIt :doc:`out_spec1D` files,
```
Thanks!

debora-pe commented on pypeit/specobjs.py:
```diff
@@ -766,6 +766,9 @@ def write_to_fits(self, subheader, outfile, overwrite=True, update_det=None,
                 header[key.upper()] = subheader[key]
+                # Also store the datetime in ISOT format
+                if key.upper() == 'MJD':
+                    header['DATETIME'] = (Time(subheader[key][0], format='mjd').isot, "Date and time of the observation in ISOT format")
```
I don't know if you noticed, but the CI tests are failing because of this. The error is:
```
                # Also store the datetime in ISOT format
                if key.upper() == 'MJD':
>                   header['DATETIME'] = (Time(subheader[key][0], format='mjd').isot, "Date and time of the observation in ISOT format")
E                   TypeError: 'float' object is not subscriptable
```

rcooke-ast replied:
Thanks @debora-pe! This line of code is fine for the reduction but not the flux calibration step. So, I've implemented a fix, and this should work again now. It would probably be a good idea to run the dev-suite just to be sure, and I've just started this now.

rcooke-ast replied:
Thanks @debora-pe! This line of code is fine for the reduction but not the flux calibration step. So, I've implemented a fix, and this should work again now. It would probably be a good idea to run the dev-suite just to be sure, and I've just started this now.

rcooke-ast committed:
fix tests

rcooke-ast committed:
no error returned

rcooke-ast commented:
Tests pass...

<img width="631" alt="image" src="https://github.com/user-attachments/assets/1974de4f-bfe6-4dcc-81f2-7e9abf67c169">

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
Merge branch 'uves_popler_tmp' into uves_popler

# Conflicts:
#	pypeit/core/flexure.py

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
unify blaze computation

Commit:
Add pseudo-blaze to test IDL files

rcooke-ast committed:
Merge branch 'develop' into uves_popler

# Conflicts:
#	pypeit/coadd1d.py
#	pypeit/sensfunc.py
#	pypeit/specobjs.py

rcooke-ast committed:
return no finecorr if fitting failure

rcooke-ast committed:
add spaces

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
Merge branch 'aat_uhrf' into uves_popler

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast committed:
replace flatfile with use_blaze

rcooke-ast committed:
fix reshape

rcooke-ast committed:
rm flatfile

rcooke-ast committed:
add checks

rcooke-ast committed:
taper the finecorr to the flatfield

rcooke-ast committed:
fix possible divide by zero

rcooke-ast committed:
remove bad/unrealistic flat model values

rcooke-ast committed:
fix for case with no flatfield

rcooke-ast commented:
OK - I've unified the approach to the blaze computation, and tests now pass. This also required a few bug fixes in the flatfield code which produces infinities in the flat model (even on the `develop` branch).
<img width="609" alt="image" src="https://github.com/user-attachments/assets/4add6046-93d9-4951-8a18-31452a19c727">

kbwestfall commented on doc/coadd1d.rst:
```diff
@@ -303,3 +303,45 @@ and launches a GUI from the `linetools`_ package. e.g.:
+require some support on how to use UVES_popler with PypeIt data, please
+join the `PypeIt Users Slack <https://pypeit-users.slack.com>`__ and
+ask for help in the #coadd1d channel. All are welcome to join using
+`this invitation link <https://join.slack.com/t/pypeit-users/shared_invite/zt-1kc4rxhsj-vKU1JnUA~8PZE~tPlu~aTg>`__.
```
I'm a little uncomfortable with this.  A few thoughts:
- Specifically, I worry about volunteering to support use of someone else's software with our output data.  I.e., users should be going to the UVES_popler team first, and only coming to us if there's a problem with our datamodel that keeps them from using their code.
- If we keep this in, we should instead have a dedicated Slack channel for UVES_popler questions; i.e., the #coadd1d channel should be specifically for use of the pypeit 1D coadd code.
- Also, instead of repeating the link here, I would refer to the community section of the landing page (README.rst) so that we don't have to maintain the invitation link in multiple places.  I.e., ":ref:`community`" somewhere, instead of the link.

rcooke-ast replied:
Sounds good. It might be something worthwhile bringing up at the next PypeIt meeting to get opinions from others. I tend to agree with you, so I've removed this, but can reinstate it if it helps. I spoke to Michael Murphy about this, and there isn't a dedicated UVES_popler Slack channel.

kbwestfall replied:
Thanks.  To be clear, I'm fine with there being a #uves_popler channel in the pypeit users slack where users can discuss that tool specifically.  I just want to keep it distinct from #coadd1d .

rcooke-ast replied:
Totally fine... but I think it's perhaps best that we don't explicitly list that they should receive help from the PypeIt Slack channel. If there's enough comments about it, then we can always add a #uver_popler channel in the Slack, but maybe best not to advertise it as such, since none of us are really able to address issues within UVES_popler. But if there's demand, I'm more than happy to add that in later, provided that Michael Murphy is happy to be the one who provides support

kbwestfall replied:
Sounds good to me!

kbwestfall commented on doc/releases/1.16.1dev.rst:
```diff
@@ -17,6 +17,7 @@ Dependency Changes
 
 Functionality/Performance Improvements and Additions
 ----------------------------------------------------
+- Added support for data coaddition with the UVES_popler GUI tool.
```
I would reword this:  "Adjusted spec1d datamodel to enable use with UVES_popler GUI tool"

rcooke-ast replied:
Yep, good call 👍 done. I also moved this to the Datamodel changes section.

kbwestfall commented on pypeit/core/skysub.py:
```diff
@@ -588,9 +588,12 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, thismask,
         If sciimg is an A-B image, then this is the global sky modeled from the A image, while `global_sky` is
         the global modeled from the A-B image. This is None if the sciimg is not an A-B image.
-    fwhmimg : `numpy.ndarray`_, None, optional:
+    fwhmimg : `numpy.ndarray`_, None, optional
```
Good catch!

kbwestfall commented on pypeit/core/telluric.py:
```diff
@@ -189,6 +189,8 @@ def read_telluric_pca(filename, wave_min=None, wave_max=None, pad_frac=0.10):
     dwave, dloglam, resln_guess, pix_per_sigma = wvutils.get_sampling(wave_grid)
     tell_pad_pix = int(np.ceil(10.0 * pix_per_sigma))
+    if dloglam == 0.0:
+        msgs.error('The telluric PCA method has zero spacing in log wavelength. This is not supported.')
```
Can we move these lines into the `wvutils.get_sampling` function?  The pro is that we don't have to repeat these lines so many time; the cons are that the error message might be less specific and there may be some application of `wvutils.get_sampling` where things can proceed, even if `dloglam==0` (although I don't know what that would be).

rcooke-ast replied:
Good idea - done!

kbwestfall commented on pypeit/core/telluric.py:
```diff
@@ -2410,18 +2419,12 @@ def __init__(self, wave, flux, ivar, gpm, telgridfile, obj_params, init_obj_mode
-        if log10_blaze_function is not None:
+        self.wave_in_arr, self.flux_in_arr, self.ivar_in_arr, self.mask_in_arr, self.log10_blaze_func_in_arr, \
+            self.nspec_in, self.norders = utils.spec_atleast_2d(
+            wave, flux, ivar, gpm, log10_blaze_function=log10_blaze_function)
```
Nice.

kbwestfall commented on pypeit/flatfield.py:
```diff
@@ -1317,10 +1317,28 @@ def fit(self, spat_illum_only=False, doqa=True, debug=False):
+                          'for slit {1:d}!'.format(winfnan[0].size, slit_spat) + msgs.newline() +
+                          'These model pixel values will be set to the raw pixel value.')
+                self.flat_model[np.where(onslit_tweak)[0][winfnan]] = rawflat[np.where(onslit_tweak)[0][winfnan]]
+                self.mspixelflat[np.where(onslit_tweak)[0][winfnan]] = 1.0
```
I feel like we should be masking these instead, but it's likely a lot of work to propagate the masking through the code...

rcooke-ast replied:
It probably is a good idea to mask, but I think these only go badly wrong in the cases when there is very little flux to actually determine the model. In the case of X-SHOOTER, not all orders fill the detector, so there's dead space with no science information, but still the model is being computed. Since we don't mask a few lines below this, I'm tempted to leave this as it is for now, and maybe have a dedicate PR for that in the future.

kbwestfall commented on pypeit/flatfield.py:
```diff
@@ -1495,7 +1522,11 @@ def spatial_fit_finecorr(self, normed, onslit_tweak, slit_idx, slit_spat, gpm, s
+
+        # If corrections are above the tolerance, then clip them to the level of the tolerance
+        illumflat_finecorr[np.where(illumflat_finecorr < 1-tolerance)] = 1-tolerance
+        illumflat_finecorr[np.where(illumflat_finecorr > 1+tolerance)] = 1+tolerance
```
Can we use `np.clip`?

rcooke-ast replied:
🤦‍♂️ of course we can... done!

kbwestfall commented on pypeit/par/pypeitpar.py:
```diff
@@ -1949,7 +1949,7 @@ class SensFuncPar(ParSet):
     see :ref:`parameters`.
     """
-    def __init__(self, flatfile=None, extrap_blu=None, extrap_red=None, samp_fact=None, multi_spec_det=None, algorithm=None, UVIS=None,
+    def __init__(self, use_blaze=None, extrap_blu=None, extrap_red=None, samp_fact=None, multi_spec_det=None, algorithm=None, UVIS=None,
```
Just logging my continued objection to the use of this term in this context...  Feel free to ignore.

rcooke-ast replied:
Fair point... I agree with you about this, and changed it throughout (and propagated these changes to the dev suite, as well).

kbwestfall commented on pypeit/pypeit.py:
```diff
@@ -1029,14 +1029,20 @@ def extract_one(self, frames, det, sciImg, bkg_redux_sciimg, objFind, initial_sk
             msgs.info(f"Extraction begins for {self.basename} on det={det}")
+            # set the flatimg, if it exists
+            try:
+                flatimg = self.caliBrate.flatimages.pixelflat_model
```
Is there a specific case that you have where this faults?  I'm confused because `pixelflat_model` is a component of the datamodel for `pypeit.flatfield.FlatImages`, so I would have thought that the attribute should always exist.

rcooke-ast replied:
It fails when a user chooses not to flatfield the data (this showed up as one of the failed dev suite tests).

debora-pe replied:
Yes, `self.caliBrate.flatimages` could be `None`. But, if it's not,  `pixelflat_model` exists (even if it's `None`). So 
`flatimg = self.caliBrate.flatimages.pixelflat_model if self.caliBrate.flatimages is not None else None` 
should work.

kbwestfall replied:
Got it.  Is there a more direct check?  Something like ``if self.caliBrate.flatimages is None``?

rcooke-ast replied:
Hmm... good point - I'm not sure, but I'll try the direct check and run the dev-suite to see if it breaks  😃

kbwestfall commented on pypeit/scripts/sensfunc.py:
```diff
@@ -66,17 +66,6 @@ def get_parser(cls, width=None):
-                                 "F|         flatfile = Calibrations/Flat_A_0_DET01.fits\n\n"
-                                 "Where Flat_A_0_DET01.fits is the flat file in your "
-                                 "Calibrations directory\n")
-
```
To maintain functionality, do we need to add a parameter that allows the user to not use the extracted flat-field spectrum during the sensitivity calculation?

rcooke-ast replied:
Since no default value is set (i.e. the default is `None`) then I think this is OK to remove. I suppose we could add in a `use_flat` boolean, but I'd rather encourage folks to put it in their `sens.par` for a record of what parameters they set.

debora-pe replied:
I would agree with @kbwestfall . I am running `sensfunc` a lot these days and I do it always from command line (I do not create a sens configuration file). I would definitely prefer if the `--use_flat` boolean parameter could exists here.

rcooke-ast replied:
OK, sounds good, and I've added it in!

kbwestfall commented on pypeit/specobjs.py:
```diff
@@ -791,6 +812,15 @@ def write_to_fits(self, subheader, outfile, overwrite=True, update_det=None,
+                        mjdval = subheader[key]
+                    else:
+                        raise ValueError('Header card must be a float or a FITS header tuple')
+                    header['DATETIME'] = (Time(mjdval, format='mjd').isot, "Date and time of the observation in ISOT format")
```
What if this is already in the header?  I.e., doesn't the header include most/all information from the raw images?  Would it make sense to instead add a new header keyword that's specific to the data reduction?  Maybe `RDXDATE` or `PYPDATE` or something similar?

rcooke-ast replied:
This is the format required by `UVES_popler`, so we need to ensure that it's in the header. I don't think it matters if it gets overwritten, but it "needs" to be there if we don't want to modify the `UVES_popler` code.

kbwestfall commented on pypeit/specobj.py:
```diff
@@ -251,11 +255,11 @@ def __init__(self, PYPELINE, DET, OBJTYPE='unknown',
     @classmethod
     def from_arrays(cls, PYPELINE:str, wave:np.ndarray, counts:np.ndarray, ivar:np.ndarray,
-                    mode='OPT', DET='DET01', SLITID=0, **kwargs):
+                    flat:np.ndarray, mode='OPT', DET='DET01', SLITID=0, **kwargs):
```
Given some of the associated changes, should `flat` be optional?

rcooke-ast replied:
You're absolutely right - good catch 👍

kbwestfall reviewed (approved):
A number of nit-picky comments from me, but hopefully they're all relatively easy to deal with.

rcooke-ast committed:
Addresses PR comments

rcooke-ast reviewed (commented):
Thanks @kbwestfall for the quick turnaround. I've addressed all comments, and made the appropriate changes to the dev suite, too. I'll await for any final comments before re-running the dev-suite to make sure I haven't broken anything in the process.

rcooke-ast committed:
explicitly check for pixelflat_model

debora-pe commented on pypeit/flatfield.py:
```diff
@@ -1317,10 +1317,28 @@ def fit(self, spat_illum_only=False, doqa=True, debug=False):
             # TODO: Add some code here to treat the edges and places where fits
             #  go bad?
 
+            # Check for infinities and NaNs
```
This is really nit-picking...but you could add these checks for infinities and NaNs + unrealistically high or low values of the model before filling `self.mspixelflat[onslit_tweak]` in line 1320, so that you will not need to set 
`self.mspixelflat[np.where(onslit_tweak)[0][winfnan]] = 1.0`
and 
`self.mspixelflat[np.where(onslit_tweak)[0][whilo]] = 1.0`

debora-pe replied:
This is unrelated, but I would like to ask you something because you understand the flat field code better than me. Here in the `fit()` method, there are several calls to `self.slits.slit_img()` and `self.slits.spatial_coordinate_image()` where the spatial flexure is a parameter, but it's never provided. While, for example, when `fit2illumflat()` is used, the spatial flexure is usually provided. 
What do you think about this? Should the spatial flexure always used?

rcooke-ast replied:
Good idea. I've moved this code up a few lines.

Regarding your query about the spatial flexure, I think it's important when computing the spatial illumination, but for the pixel-flat it's less critical since a 2D fit is performed and the entire slit is normalised. @jhennawi may have a more well-informed idea, as I'm pretty sure he wrote this code.

This is quite intricate as well, since you always want to compute the spatial shift relative to the `trace` frame (which defines the slit edges), but the edges then get tweaked by the `illumflat`. So, I *think* that all of the spatial flexure corrections are done relative to the `illumflat`. Note that the spatial flexure correction is still somewhat primitive, but the correction is good enough for most applications.

debora-pe replied:
Thanks for your answer @rcooke-ast. It is a bit more clear now and indeed I hadn't thought how the tweaking of the slits affect the spatial flexure.

debora-pe commented on pypeit/flatfield.py:
```diff
@@ -1317,10 +1317,28 @@ def fit(self, spat_illum_only=False, doqa=True, debug=False):
+            whilo = np.where((self.flat_model[onslit_tweak] >= nonlinear_counts) |
+                             (self.flat_model[onslit_tweak] <= 0.0))
+            if whilo[0].size != 0:
+                msgs.warn('There are {0:d} pixels with non-finite values in the flat-field model '
```
I think this warning should say something like `''There are {0:d} pixels with unrealistically high or low values in the flat-field model ..."`

rcooke-ast replied:
Well-spotted - updated 👍

debora-pe reviewed (approved):
Thanks @rcooke-ast for this and for fixing the bugs. I have a few comments (one kind of a philosophical question) and a request for the `sensfunc` script.
I am approving anyway.

rcooke-ast reviewed (commented):
Comments addressed - thanks @debora-pe!

rcooke-ast committed:
added use_flat as a keyword argument

rcooke-ast committed:
flatfield model masking

rcooke-ast committed:
fix argument value

debora-pe commented on pypeit/scripts/sensfunc.py:
```diff
@@ -111,13 +110,14 @@ def main(args):
-        if args.flatfile is not None and args.sens_file is not None:
-            msgs.error("It is not possible to set --flatfile and simultaneously use a .sens "
+
+        if not args.use_flat and args.sens_file is not None:
```
Thank you for adding this.
I think here you want `if args.use_flat and args.sens_file is not None:`
The idea is: if the `--use_flat` flag is not used, it means `args.use_flat` is `False` and in that case we use the default value for the instrument or the value in the `.sens` file. If the `--use_flat` flag  is used, `args.use_flat` is `True` and that would be the value used, which clashes with the `.sens` file (if provided), hence this error.

rcooke-ast replied:
Indeed -- good catch! I'll fix this and restart my dev-suite run. Thanks for catching that!

rcooke-ast committed:
fix argument value

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast commented:
Tests pass...
<img width="620" alt="image" src="https://github.com/user-attachments/assets/17f71c2c-9b0c-423a-9f90-69b10244b3cf">

rcooke-ast committed:
Merge branch 'develop' into uves_popler

rcooke-ast commented:
Tests still pass
<img width="620" alt="image" src="https://github.com/user-attachments/assets/2611a1e8-8cbb-4dc2-b08a-db8c27561114">

rcooke-ast commented:
I've checked this branch works with `UVES_popler`, so I'm merging.
